reasoner/planner/AnswerCountEstimator.java

```diff
@@ -60,210 +59,195 @@
+            conjunctionModels.get(conjunction).conjunctionSummary.resolvables().forEach(incrementalEstimator::extend);
+            fullConjunctionEstimators.put(conjunction, incrementalEstimator);
+        }
+        IncrementalEstimator estimator = fullConjunctionEstimators.get(conjunction);
+        Set<Variable> estimateableVariables = ReasonerPlanner.estimateableVariables(variables);
+        return estimator.scaledEstimate(estimateableVariables);
     }
 
-    private static class ConjunctionConcludableStack {
-
-        ArrayList<ResolvableConjunction> stack;
-        Map<ResolvableConjunction, Concludable> concludables;
-
-        ConjunctionConcludableStack() {
-            this.stack = new ArrayList<>();
-            this.concludables = new HashMap<>();
-        }
-
-        private void add(ResolvableConjunction conjunction) {
-            stack.add(conjunction);
-            concludables.put(conjunction, null);
-        }
-
-        private ResolvableConjunction last() {
-            return stack.get(stack.size() - 1);
-        }
+    public IncrementalEstimator createIncrementalEstimator(ResolvableConjunction conjunction) {
+        if (!conjunctionModels.containsKey(conjunction)) buildConjunctionModel(conjunction);
+        return new IncrementalEstimator(conjunctionModels.get(conjunction));
+    }
 
-        private boolean contains(ResolvableConjunction conjunction) {
-            return concludables.containsKey(conjunction);
-        }
+    public static class IncrementalEstimator {
+        private final ConjunctionModel conjunctionModel;
+        private final Map<LocalModel, Pair<Double, Optional<Variable>>> modelScale;
+        private final Map<Variable, Double> bestUnary;
+        private final Map<Variable, Set<LocalModel>> modelsWithVar;
 
-        private void pop() {
-            ResolvableConjunction conjunction = stack.remove(stack.size() - 1);
-            concludables.remove(conjunction);
+        private IncrementalEstimator(ConjunctionModel conjunctionModel) {
+            this.conjunctionModel = conjunctionModel;
+            this.modelScale = new HashMap<>();
+            this.bestUnary = new HashMap<>();
+            this.modelsWithVar = new HashMap<>();
         }
 
-        private void setConcludable(ResolvableConjunction conjunction, Concludable concludable) {
-            assert stack.contains(conjunction);
-            concludables.put(conjunction, concludable);
-        }
+        public void extend(Resolvable<?> resolvable) {
+            Map<Variable, Double> unaryUpdates = new HashMap<>();
+            List<LocalModel> models = conjunctionModel.modelsForResolvable(resolvable);
+            for (LocalModel model : models) {
+                addAndCollectUnaryUpdates(model, unaryUpdates);
+            }
+            propagate(unaryUpdates);
+        }
+
+        private void addAndCollectUnaryUpdates(LocalModel model, Map<Variable, Double> unaryUpdates) {
```

Comment:
When a model is 'added' its estimate for each variable is compared to the best-known estimate for each variable to determine the scaling-factor is needed. The scaling factor is the ratio of best-known estimate to this model's estimate (at most 1).

The scaling-factor is then used to scale down this model's estimates for each variable - If any of these are lower they are collected to be propagated later.